src/D3MHub.sol

```diff
@@ -0,0 +1,601 @@
+                unchecked {
+                    toUnwind = _max(toUnwind, currentAssets - targetAssets);
+                }
+            }
+
+            if (toUnwind > 0) {
+                _unwind(
+                    ilk,
+                    _pool,
+                    toUnwind,
+                    Mode.NORMAL,
+                    currentAssets
+                );
+            } else {
+                uint256 toWind;
+                // All the subtractions are safe as otherwise toUnwind is > 0
+                unchecked {
+                    toWind = _min(
+                                _min(
+                                    _min(
+                                        targetAssets - currentAssets,
+                                        lineWad - Art
+                                    ),
+                                    (Line - debt) / RAY
+                                ),
+                                _pool.maxDeposit() // Determine if the pool limits our total deposits
+                             );
+                }
+                require(Art + toWind <= MAXINT256, "D3MHub/wind-overflow");
+                _wind(ilk, _pool, toWind);
+            }
+        }
+
+        _pool.postDebtChange("exec");
+    }
+
+    /**
+        @notice Collect interest and send to vow.
+        Total collected will be constrained by the number of assets available to
+        be withdrawn from the pool.
+        @param ilk bytes32 of the D3M ilk name
+    */
+    function reap(bytes32 ilk) external lock {
+        ID3MPool _pool = ilks[ilk].pool;
+
+        require(vat.live() == 1, "D3MHub/no-reap-during-shutdown");
+        require(ilks[ilk].tic == 0, "D3MHub/pool-not-live");
+        require(ilks[ilk].plan.active(), "D3MHub/plan-not-active");
+
+        _pool.preDebtChange("reap");
+        uint256 assetBalance = _pool.assetBalance();
+        (, uint256 daiDebt) = vat.urns(ilk, address(_pool));
+        if (assetBalance > daiDebt) {
+            uint256 fees;
+            unchecked {
+                fees = assetBalance - daiDebt;
+            }
+            fees = _min(fees, _pool.maxWithdraw());
+            _pool.withdraw(fees);
+            daiJoin.join(vow, fees);
+            emit Reap(ilk, fees);
+        }
+        _pool.postDebtChange("reap");
+    }
+
+    /**
+        @notice Allow Users to return vat gem for Pool Shares.
+        This will only occur during Global Settlement when users receive
+        collateral for their Dai.
+        @param ilk bytes32 of the D3M ilk name
+        @param usr address that should receive the shares from the pool
+        @param wad amount of gems that the msg.sender is returning
+    */
+    function exit(bytes32 ilk, address usr, uint256 wad) external lock {
+        require(wad <= MAXINT256, "D3MHub/overflow");
+        vat.slip(ilk, msg.sender, -int256(wad));
+        ID3MPool _pool = ilks[ilk].pool;
+        _pool.transfer(usr, wad);
```

Comment:
Pool doesn't need to be cached as it is only used once.

Reply:
0b42cfa